Configure Xdebug debugging sessions to be started by a trigger by default, fixes #3593, fixes #3731

nico-loeber opened the pull request:
## The Problem/Issue/Bug:

* #3593 
* #3731 

As an experienced PHP developer 
- I would like to be able to use all of Xdebug's features with zero configuration.
- I would like to use the development features without triggering a debugging session.
- I do not expect any errors or warnings when executing PHP scripts on the command line.
- I would like to be able to generate line and path coverage reports of my code using phpunit and xdebug.
- I expect to be able to start debugging session via the XDebug browser extension or from my IDE.

As a non experienced PHP developer 
- I would like to start a xdebug debugging session with minimal effort.
- I would like to use the improved error traces provided by xdebug.

## How this PR solves the problem:

This PR is based on the results of the issue #3731.

Xdebug is enabled by default with all of its features. A debugging session is no longer started automatically. 
This allows the usage of the Xdebug development features during development of your application (and not debugging it) as well as triggering alternative features like profiling or tracing.
In this mode you can enable debugging sessions by cookies (browser extension or bookmarks) or by environment variables on the cli (set by your IDE).
To enable the automatic start of a debugging session you can run `ddev xdebug autostart` or by setting the value `xdebug_autostart: true` in the .ddev/config.yml. 
This will start a debugging session on each request and on every started script on the command line. This is the current default behavior, when enabling xdebug.
The Xdebug extension can still be disabled by running `ddev xdebug off` or setting `xdebug: false` in the configuration.

### Architectural Decisions

#### Preferring "autostart" over "trigger" as a new mode
As discussed in #3731 the initial idea was to introduce a new mode "trigger" to provide new behavior to stay compatible to the current default behavior. 
I did some research and asked developers what they would expect would happen when they execute "ddev xdebug trigger", in most cases they could not imagine the concrete behavior or describe it precise. 
At the same time "ddev xdebug autostart" was easy to understand and describe for everyone.
The current default behavior of xdebug in ddev was also not expected by the asked developers (no ddev users).

#### Enabling Xdebug by default
Enabling Xdebug is a reasonable default as it provides useful development features like better var_dump function and traces for errors. 
Using Xdebug in a development environment is a common default. To test your application in a more production like environment you can still disable the extension.
With disabled autostart of debugging sessions there are no drawbacks in running Xdebug during development. 

#### Disabling autostart mode by default
Automatically starting debugging sessions is not always intended. 
There are serveral common situations in which you do not like to start a debugging session when Xdebug is active:
- Debugging your Application with xdebug extended var_dump and xdebug error traces (Of course NOBODY would do that)
- Generating line or path coverage reports on the command line
- Debugging a single request, controlled by your browser plugin instead of debuggin all requests
- Using the profiling ot tracing features of xdebug
- You like to debug requests in the browser but not on the command line
As debugging all requests on CLI and in browser seems to me a special use case this should not be the default.

#### Using phpenmod for switchting the configuration
phpenmod is a tool to enable or disable php modules. It is not meant for switching configurations. 
I decided to use it anyway, as it makes it easy to apply the configuration change for all php variants (Module, FPM and CLI) and all versions (5.6 - 8.1). 
Also the current xdebug configuration is already placed in a mods-available configuration, so using those files for configuration changes seems to be acceptable in the DDEV project.

#### Supporting Xdebug 2
As discussed in #3731 DDEV does not implement new features for xdebug 2.
However this pull request does support xdebug 2, because not supporting it would have added complexity to this pull request. 
Omitting support for xdebug 2 would cause errors executing `phpenmod xdebug_autostart` as the auto start configuration file would not exist for php version lower than 7.2. Also the behavior between xdebug 2 and xdebug 3 php versions would change, which would add complexity for the user and to the documentation. 

### Breaking Changes
Xdebug will be enabled by default. add `xdebug_enabled: false` to disable it.
Debugging session will not start automatically. You can trigger them by browser or environment variable on CLI, or you have to run `ddev xdebug autostart` to debug every request, this will equal the current behavior when xdebug is enabled.


## Manual Testing Instructions:

- Run `ddev config` in a new project
- When using an existing project, make sure there is no `xdebug_enabled: false` in `.ddev/config.yml`

### Default behavior
- Run `ddev xdebug status`
  - should return "xdebug enabled (debug,profile,trace)"
- Browse your website 
  - no debugging session should start
- Open your browser and start a debugging session by using a bookmark or the xdebug extension
  - a debugging session should start
- Run a debugging session on cli with your IDE
  - a debugging session should start
- Run any php script on cli
  - no debugging session should start and there should be no unexpected xdebug output

### Switching Mode
- Run `ddev xdebug autostart`
- Run `ddev xdebug status`
  - should return "xdebug enabled with autostart (debug,develop)"
- Open your application in browser
  - a debugging session should start
- Run any script on cli
  - a debugging session should start
- Run `ddev xdebug trigger`
  -  See default behavior

## Automated Testing Overview:
- xdebug_test.go
- ddevapp_test.go

## Related Issue Link(s):
- https://github.com/drud/ddev/issues/3731

## Release/Deployment notes:
N/A

## Open todos
- [x] Implement changes
- [x] Update Container
- [x] Update / add test caes
- [x] Update documentation

## Related

#### Xdebug bookmark generator
https://www.jetbrains.com/phpstorm/marklets/

#### Xdebug Helper Browser Extension
https://chrome.google.com/webstore/detail/xdebug-helper/eadndfjplgieldjbigjakmdgkmoaaaoc?hl=de

#### PHPStorm Documentation on debugging
https://www.jetbrains.com/help/phpstorm/zero-configuration-debugging.html#activate-debugger-on-server

#### Xdebug error traces
![image](https://user-images.githubusercontent.com/18308418/162589306-5765b1cb-1a6c-4157-984b-045b1cf3ae37.png)

#### Xdebug improved var_dumps
![image](https://user-images.githubusercontent.com/18308418/162589283-7c7b0ca4-b045-482d-b160-5aa22930fc5b.png)


<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3754"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

nico-loeber commented:
> You still have this as draft...

I did not fix the tests as I was expecting some feedback from you, @rfay .

Also I am not sure how to handle the changes to the container. I read the docs and saw that I need to push the modified container to an official docker registry and add its name to same configuration.
But I guess I should not commit this configuration change as it shouldn't be merged. So I am not sure how to get the build pipeline fixed.

So, if you tell me to continue and give me some hint on how to continue, I will do.

rfay commented:
That's all fine! I'll take a look then when I get home. I usually wait until something is out of draft before looking.

rfay reviewed (changes requested):
I don't object to completely removing the config.yaml `xdebug_enabled`. It's long been obsolete and unuseful... and just lived on.

What I'd rather do is to keep the default behavior the same as it is today, which I think is possible with your approach. The default behavior has gotten many people started who never were able to use xdebug before. I've always been suspicious and seen lots of confusion around the browser-extension techniques. 

But it seems to me like you're on track to being able to do that just fine here (leaving the default behavior the same). And then `ddev xdebug trigger` could turn on the mode you like. 

It's also possible to add a little config to set the trigger mode, although I like the command better, and those that like it could have it as a post-start command.

Thanks so much for the initiative on this, and sorry I've been so slow to review. I think it has plenty of potential, and I'm happy to help. It's easy to push images when you're ready, and you can actually push and use your own (temporarily) by editing pkg/version/version.go if you prefer.

nico-loeber commented:
I think we need to define "changing the default behavior" more precise.
If this means "If I open the website  in a browser and I did not change any setting, I don't want any debug session to start" then the default behavior is not affected by these changes. There is also no notable performance impact.

Enabling xdebug in development mode by default, which is currently proposed by this pull request, will provide you better error reporting, infinite recursion detection and formatted var_dumps. Also additional features can be triggert by external tools.
These are features I would expect from my development environment as a php developer.
These features can still be disabled by running `ddev xdebug off`.
This is also required if you like to use phpstorm to debug phpunit tests.

In case you agree on this I will continue to fix the tests.

rfay commented:
I used the wrong terms when I said "changing behavior". I should have said "changing default user experience". I want the outstanding simple default user experience to remain the same, with all the extra trigger behavior (and other xdebug features) you want added on.

nico-loeber commented:
Due to the fact that there are problems with Typo3 projects when running them with xdebug development mode enabled (Failing ddev tests because of error output as well as problems with high max_function_nesting_level) I decided to only enable xdebug modes that are enabled by a trigger. So the default behavior will not change at all as long as no trigger is sent with the request.

nico-loeber commented:
I have added a retry mechanism to the xdebug tests. If the xdebug connection is not established after the first HTTP request, there are up to three more.
I hope this will make these tests more predictable.

nico-loeber commented:
I've updated the documentation to match the new behavior of xdebug better. 
I also tried to make the documentation a bit more "use case focused" so I tried to place the "how to debug" part at the beginning.
I did add some information on how to start a debugging session on the command line without starting it via IDE.
I did remove the documentation on how to configure a Run Configuration as it is PHPStorm specific and well documented here: https://www.jetbrains.com/help/phpstorm/debugging-a-php-cli-script.html#create-a-run-debug-configuration-for-a-php-script-manually

rfay commented:
Thanks, I'll make time to spend some time with it. I'm an old hand at doing a "run configuration", but of course... most people aren't successful with that. You used to have to do that before PhpStorm got so good at the zero-config.

rfay commented:
Rebased and pacified spellchecker

rfay commented on pkg/versionconstants/versionconstants.go:
```diff
@@ -14,10 +14,10 @@ var DdevVersion = "v0.0.0-overridden-by-make" // Note that this is overridden by
 
 // WebImg defines the default web image used for applications.
-var WebImg = "drud/ddev-webserver"
+var WebImg = "niconator/ddev-webserver"
```
Let's remember to push drud/ddev-webserver before this gets pulled.

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
+1. Toggle the Start Listen PHP Debug Connections button start listening php debug connections on the PhpStorm toolbar.
+   ![Start listening to debugging connections](images/ps_quick_start_zero_debug_stop_listening_debug_connections_toolbar.png)
+2. Set a breakpoint
+3. When you start debugging a [website](#Debug a website) or [CLI application](#Debug a CLI application) PhpStorm might
```
These links don't show up right in markdown. 

![image](https://user-images.githubusercontent.com/112444/174677996-5b79ab1e-af1a-4024-bf2e-8a6039157fdf.png)

See https://github.com/drud/ddev/blob/d2ed0c21c89e1f8f9aabdb4c1bee98993d1c522e/docs/users/step-debugging.md

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
 
-Note when using this recommended option:
+To start and stop Xdebug from the browser, you need to set a cookie parameter. You can use one of the
+[available browser debugging extensions](#Available browser debugging extensions). Alternatively you can debug every
```
This links don't show in markdown correctly, see https://github.com/drud/ddev/blob/d2ed0c21c89e1f8f9aabdb4c1bee98993d1c522e/docs/users/step-debugging.md

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
-4. Add file mappings for the files on the server. Click on the local repo path and add "/var/www/html" as the "Absolute path on the server" and your repository root as the path on the host.
-5. Set an appropriate breakpoint.
-6. Start debugging by clicking the "debug" button, which will launch a page in your browser.
+| Chrome                                                                                        | Firefox                                                                                                                                                                        | Internet Explorer                                                               | Microsoft Edge                             ...
```
At this point IE is just fluff I think. You're the Windows expert so I defer to you, but consider removing.

nico-loeber replied:
Support was dropped on june 15th 2022. Just a few days ago - finally. I agree, will remove it.

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
+### Debug a command line application
 
-Server creation:
+To start a step debug session for a PHP script on the command line prefix your command line
```
```suggestion
To start a step debug session for a PHP script on the command line, prefix your command line
```

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
 
-Server creation:
+To start a step debug session for a PHP script on the command line prefix your command line
+with `php -dxdebug.start_with_request=yes {path_to_your_script}` when executing the script inside the ddev container.
```
```suggestion
with `php -dxdebug.start_with_request=yes {path_to_your_script}` when executing the script inside the ddev container or with `ddev exec`.
```

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
-6. In a browser, visit your project, you should hit the breakpoint.
+* You can let xdebug start a debugging session for every request with `ddev xdebug autostart` and
+  return to the default with `ddev xdebug trigger`
+* You can disable xdebug with `ddev xdebug off` or setting `xdebug_enabled: false` in your `.ddev/config.yml`
```
Consider removing xdebug_enabled from app/yaml

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
+  return to the default with `ddev xdebug trigger`
+* You can disable xdebug with `ddev xdebug off` or setting `xdebug_enabled: false` in your `.ddev/config.yml`
+* `ddev xdebug status` will show you the current xdebug configuration
+* If xdebug is disabled in your configuration, you can enable it for a already running container with `ddev xdebug on`
```
```suggestion
* If xdebug is disabled in your configuration, you can enable it for a already-running project with `ddev xdebug on`
```

rfay commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
-By default, ddev is set up to contact the default port, port 9003 on your IDE. However, if you have something else listening on that port or your IDE does not yet default to 9003, you'll need to change the port. (PhpStorm and vscode have switch to supporting 9003 instead of 9000 for some time now.)
+By default, ddev is set up to contact the default port, port 9003 on your IDE. However, if you have something else
+listening on that port or your IDE does not yet default to 9003, you'll need to change the port. (PhpStorm and vscode
+have switch to supporting 9003 instead of 9000 for some time now.)
```
```suggestion
have switched to supporting 9003 instead of 9000 for some time now.)
```

rfay commented on docs/users/step-debugging.md:
```diff
@@ -93,26 +93,61 @@ xdebug.client_port=9000
+configure it on your workstation.
+
+The basic thing to understand about xdebug is that it's a network protocol. Your IDE (like PhpStorm) will listen on the
+xdebug port (9003 by default in v1.19+, previously 9000). Then if a xdebug session is started, php inside the container
```
```suggestion
xdebug port (9003 by default in v1.19+, previously 9000). When an xdebug session is started, php inside the container
```

rfay commented on docs/users/step-debugging.md:
```diff
@@ -93,26 +93,61 @@ xdebug.client_port=9000
+The basic thing to understand about xdebug is that it's a network protocol. Your IDE (like PhpStorm) will listen on the
+xdebug port (9003 by default in v1.19+, previously 9000). Then if a xdebug session is started, php inside the container
+will try to open a TCP connection to the IDE. When starting the session via a browser xdebug will try to connect back to
+client, otherwise `host.docker.internal:9003` will be used to connect, which is provided by docker. So you have to make
```
```suggestion
client, otherwise `host.docker.internal:9003` will be used to connect, which is provided by DDEV or Docker Desktop. So you have to make
```

nico-loeber replied:
host.docker.internal is not set when using docker on linux?

rfay replied:
It's set by ddev by exploring the network topology on Linux or Colima. Only Docker Desktop sets it otherwise. Ddev generates docker-compose stanza to define it, and figures out what it should be. Not perfect either.

rfay reviewed (changes requested):
This first pass is just looking at the code and reading the doc, haven't done manual testing yet. A few items for discussion and some random docs suggestions. The biggest item for discussion is whether we should pull xdebug_enabled out of app and config.yaml.

nico-loeber commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -749,124 +749,207 @@ func TestDdevXdebugEnabled(t *testing.T) {
 		assert.NoError(err)
 		err = os.RemoveAll(projDir)
 		assert.NoError(err)
+		_ = listener.Close()
```
As you might see in the history it was err := with an assert for no error. Sometimes the listener is already closed (Some sort of garbage collection?).

nico-loeber commented:
Hmm, I'm not sure if you should remove the xdebug setting from config.yml.
I can't see any reason why it should be turned off for a development environment...
I know there are some PHP extensions that can conflict with XDebug, such as Ioncube (source code encryption).

My suggestion would be to keep the setting for now, and if feedback from the community shows that no one is using it anymore, we could remove it at a later date. What do you think, @rfay?

PS: Sorry for the "Update docs/users/step-debugging.md" commits, just clicked the commit button and didn't see the "add to batch" feature.

rfay reviewed (changes requested):
- [ ]  ddev xdebug needs proper help (and other things?). (It looks like you forgot to add the xdebug script to the repo, please add it.)
- [ ]  AFAICT, `ddev xdebug on` doesn’t work any more with basic phpstorm listener
- [ ]  Explain the meaning of trigger and autostart. Is autostart intended to be the same as “on”?
- [ ]  [[cli-usage.md](http://cli-usage.md/)](http://cli-usage.md) needs update
- [ ]  config_yaml.md needs update
- [ ]  [[phpstorm.md](http://phpstorm.md/)](http://phpstorm.md) needs update
- [ ]  [[xdebug-profiling.md](http://xdebug-profiling.md/)](http://xdebug-profiling.md) may need update?
- [ ] Behavior of `xdebug_enabled` needs to be explained in config.yaml comments (templates.go)

rfay commented:
I do find the xdebug that you wanted is checked in... but oddly I only see the behavior of the old script. Looking everywhere for something that's doing that.

rfay commented on pkg/ddevapp/global_dotddev_assets/commands/web/xdebug:
```diff
@@ -17,15 +17,25 @@ case $1 in
     on|true|enable)
```
Usage statement (above) needs to be updated

rfay commented:
If you want to keep the `xdebug_enabled` config in app/config.yaml, shouldn't it take the values on/off/trigger/autostart ?

nico-loeber commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
-**If you are using PhpStorm inside WSL2 (or perhaps other Linux configurations), under `Help→ Edit Custom VM Options`, add an additional line: `-Djava.net.preferIPv4Stack=true` This makes PhpStorm listen for Xdebug using IPV4; the Linux version of PhpStorm seems to default to using only IPV6.**
-
-#### PhpStorm Zero-Configuration Debugging
+1. Install the [php-debug](https://marketplace.visualstudio.com/items?itemName=felixfbecker.php-debug) extension.
```
I'm not the author of the vscode part. So no, it is not tested by me. I just reordered the existing documentation and updated it accodingly to the new xdebug behavior.

nico-loeber commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
+### Debug a website
 
-Note when using this recommended option:
+To start and stop Xdebug from the browser, you need to set a cookie parameter. You can use one of the
```
You can start debugging sessions without any configuration now. If you like xdebug sessions to start automatically you have to run `ddev xdebug autostart`.

rfay replied:
That's not my experience. You have to `ddev debug autostart` to do it with no configuration.

Or maybe it's because all existing projects have `xdebug_enabled: false` in them?

nico-loeber replied:
> Or maybe it's because all existing projects have xdebug_enabled: false in them?

Yes that's proabably the reason. I noted this in the manual testing instructions.
Of course we should not change the behavior for projects that already have `xdebug_enabled: false` set explicitly. Maybe this was done by intend.
For new project `xdebug_enabled` should not appear in the config.yaml at all when running `ddev config` and the default will be `on`. As you already noticed the "new on" won't have any negative effect.

rfay replied:
All projects in the world have `xdebug_enabled: false`, because it's been the default since the beginning (before `ddev xdebug` even existed, when you had to change this setting to debug).  So we have to figure our way around this bit of friction.

nico-loeber commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
-4. In the menu, choose Run->Start Debugging.You may have to select "Listen for XDebug" by the green arrowhead at the top left. The bottom pane of vscode should now be orange (live) and should say "Listen for XDebug".
-5. Enable XDebug with `ddev xdebug on`
-6. In a browser, visit your project, you should hit the breakpoint.
+* You can let xdebug start a debugging session for every request with `ddev xdebug autostart` and
```
`ddev xdebug autostart` will also start debugging sessions on cli, not only web requests... one of the reasons for the change of the behavior as this caused trouble.

rfay replied:
I'll have to go back and test, but I think the problem here is that you changed the default value for xdebug_enabled, but of course there are thousands of projects that have it set to false. So this will currently work (without intervention) only on new projects.

nico-loeber replied:
As `ddev xdebug autostart` will also run `ddev xdebug on` this shouldn't affect this specific use case here. I will accept your suggestions adding "web" but maybe we should add another line that `xdebug autostart` will not only affect web requests but also cli scripts.

nico-loeber commented on docs/users/step-debugging.md:
```diff
@@ -1,90 +1,90 @@
+  return to the default with `ddev xdebug trigger`
+* You can disable xdebug with `ddev xdebug off` or setting `xdebug_enabled: false` in your `.ddev/config.yml`
+* `ddev xdebug status` will show you the current xdebug configuration
+* If xdebug is disabled in your configuration, you can enable it for a already-running project with `ddev xdebug on`
```
`ddev xdebug autostart` and `ddev xdebug trigger` is explained in the first line. And no, I would not suggest using using xdebug autostart, as it will cause problems if your website executes AJAX Requests, or contains images served by php or CLI scripts run in parallel while you are debugging.
`ddev xdebug autostart` is just a "compatbility" mode for the current `xdebug on` behavior and might be useful for people who do not use an IDE for development.

nico-loeber commented:
> ddev debug autostart seems to have the previous behavior of ddev xdebug on, but ddev xdebug on doesn't seem to do anything any more.

`ddev xdebug off` will unload the xdebug extension, you won't be able to create test coverage reports, or do any debugging.
`ddev xdebug on` will load the xdebug extension. You are now able to create coverage reports for your tests and you can start debugging sessions if you send a trigger. 
`ddev xdebug autostart` is the current "on" behavior, that will automatically trigger a xdebug session for every single web requests and php executing on the command line.

> If you want to keep the xdebug_enabled config in app/config.yaml, shouldn't it take the values on/off/trigger/autostart ?

xdebug_enabled should only affect if the xdebug extension is available. I thought about such a switch and decided against it.
As you might see in my commit history I added an xdebug_autostart: true/false switch instead. But when working with my fork I realized that no one should ever commit xdebug_autostart: true to a project as this would cause trouble for other people working on the same project, as they would probably not expect debugging sessions to automatically start when they work on the project.
So I removed this setting again for simplicity and to avoid the problems that are caused by automatically started debugging sessions.

Debugging sessions might trigger anti virus / firewall alerts, will trigger errors in output and logs if there is no listening port on your pc and if there is a listening port on your pc they will block the executing of the request in the browser / on the command line. So they should not start unintended.

rfay commented:
As mentioned earlier... I think the problem is that this doesn't work as you expect on existing project that have `xdebug_enabled: false` in them. I'm sure we can figure out what to do about that or how to doc around it.

nico-loeber commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -749,124 +749,207 @@ func TestDdevXdebugEnabled(t *testing.T) {
 		// Run with xdebug enabled
 		_, _, err = app.Exec(&ddevapp.ExecOpts{
-			Cmd: "enable_xdebug",
+			Cmd: "enable_xdebug_autostart",
```
I would also like to see the bash scripts removed in this workflow, as bash code is hard to test and adds an extra layer.
But please let's consider this as an improvement we can care about later.

rfay replied:
I was saying that these separate bash scripts should be incorporated into the xdebug bash script (the custom command)

nico-loeber commented:
> As mentioned earlier... I think the problem is that this doesn't work as you expect on existing project that have xdebug_enabled: false in them. I'm sure we can figure out what to do about that or how to doc around it.

I'm not sure if this is a big problem. Let's review the use cases...

Given I have an project with xdebug_enabled: false and I am not aware of this as it was automatically added...
- In case I do not want to use any xdebug feature everything is fine
- In case I expect to be able to start debugging sessions by a browser plugin as suggested by the new documentation I might be confused
- In case I run `xdebug on` I might expect a debugging session to strart automatically as I am not aware of the changed behavior.

Given I have an project with xdebug_enabled: false and I am aware of this as I configured this manually...
- I am aware that xdebug is disabled and I did this for reason. I do not expect it to be enabled after a ddev update.

So yes I do see a small problem for users that are not aware that xdebug is disabled for existing projects or that are not aware of the changed behavior.
We could add a notice to the xdebug documentation informing about the changed default behavior in ddev version 1.x.0.
Furthermore, this should be clearly noted in the release notes.
This is indeed a breaking change but I think most developers that regulary use step debugging will welcome it. We just need to cleary communicate this change.

rfay commented:
> So yes I do see a small problem for users that are not aware that xdebug is disabled for existing projects or that are not aware of the changed behavior.

It's not really a small problem. It's "It doesn't work at all". All existing projects have it disabled, and when they do `ddev xdebug on` it doesn't get them working at all. That's what happened to me.

Lots of people would come to Discord and get their problem solved quickly... but only 5% or so of people with a problem come to do that. But every single user that uses xdebug, on every single existing project, will be affected currently, so we have to figure out how to sort this out.

nico-loeber commented:
> It's not really a small problem. It's "It doesn't work at all". All existing projects have it disabled, and when they do ddev xdebug on it doesn't get them working at all. That's what happened to me.

`ddev xdebug autostart` should still work, even `xdebug_enabled: false` is set.
`ddev xdebug on` should also work if you use a trigger (browser plugin or IDE).

If this does not work, it's probably a bug. Maybe I just misunderstand you, let's have a chat or maybe a short call about that tomorrow.

rfay commented:
Sounds good.

rfay commented:
Before we talk, just try it with a traditional .ddev/config.yaml, that has xdebug_enabled=false in it.

rfay commented:
Rebased, but I didn't push new image since it's still on nicolator. Only thing that should have changed in image since your push is your PhpStorm coverage directory fix.

rfay commented:
We should get some version of this, perhaps with a little bit of change, into v1.20.0, which will be coming up soon. If you could rebase (docs will have to be transplanted) and re-push your image after rebasing, we should be on the way. I'm pretty sure we can come up with a good compromise on the user experience, and I'll spend a little time with this again after you rebase. Thanks for all the effort on this!

rfay commented:
Are you interested in continuing with this @nico-loeber ? It seems like a worthy project, and I'm quite sure we can do it without jettisoning the current experience. 

Right now, the docs are pretty messed up for rebase, and will have to be rejiggered I think. 

One option that is certainly out there is to do a simple PR that just adds the behavior you like as an option, rather than having to change everything for everybody.

rfay commented:
I'd like to close this for now, just because the merge conflic is now quite complex and because we have a clearer idea how to configure it without changing the existing behavior. 

I'm quite sure that this will be a welcome feature, and I think we can do it in a very straightforward way. I totally appreciate how much effort you put into both the PR and educating me on its value.

Closing for now; if you want to rebase and continue on it I'll happily reopen, of course the wintertime with extra hours and no flying will come soon!
